src/Avalonia.Controls/Primitives/UniformGrid.cs

```diff
@@ -89,27 +89,45 @@ protected override Size MeasureOverride(Size availableSize)
 
         protected override Size ArrangeOverride(Size finalSize)
         {
-            var x = FirstColumn;
-            var y = 0;
-
-            var width = finalSize.Width / _columns;
-            var height = finalSize.Height / _rows;
+            var columnIndex = FirstColumn;
+            var columnWidth = finalSize.Width / _columns;
+            var rowIndex = 0;
+            var rowHeight = finalSize.Height / _rows;
+            var x = 0.0;
+            var y = 0.0;
+            var nextY = 0.0;
 
             foreach (var child in Children)
             {
-                if (!child.IsVisible)
+                if (child.IsVisible)
                 {
-                    continue;
-                }
+                    // Layout scaling may cause the child to take a different size than the one
+                    // requested. Layout each each child with it's top-left aligned against the
+                    // previous column/row bounds and request the bottom-right to be placed in
+                    // the ideal position.
+                    var topLeft = new Point(x, y);
```

Comment:
We can't rely on the previous child's bounds here: it might effectively be completely different (try any child with non-stretch alignment) from the cell's size, breaking the uniformity.

Example:

```xaml
<UniformGrid Columns="2">
  <Border Width="100" Height="100" Background="Blue" />
  <Border Width="100" Height="100" Background="Gray" />
  <Border Width="50" Height="100" Background="Blue" HorizontalAlignment="Left" />
  <Border Width="50" Height="100" Background="Gray" HorizontalAlignment="Left" />
</UniformGrid>
```

It might even be out of bounds of the `UniformGrid` (if the children are larger than their container), causing an invalid rectangle to be passed to `Arrange`, and an `InvalidOperationException`.

Example:
```xaml
<UniformGrid Width="50" Height="50">
  <Border Width="100" Height="100" />
  <Border Width="100" Height="100" />
</UniformGrid>
```